File: 0701-0800/0746.min-cost-climbing-stairs.cpp
// TODO: 尝试一下题解里的DP, 返回值为一个变量，和自己的公式有什么不同
#include "common.h"

class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int two = 0, one = 0;
        for (int x : cost) {
            /*
                new_one = min(one, two) + x;
                new_two = one;
             */
             int old = one;
             one = min(one, two) + x;
             two = old;
        }
        return min(one, two);
    }
};

int main() {
  vector<int> cost;
  INPUT_ARRAY(cost);
  cout << Solution().minCostClimbingStairs(cost) << endl;
  return 0;
}

File: 0701-0800/0746-min-cost-climbing-stairs.py
from typing import List

class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        if len(cost) < 2:
            return 0
        two = cost[0]
        one = cost[1]
        for i in range(2, len(cost)):
            old_one = one
            one = min(two, one) + cost[i]
            two = old_one
        return min(one, two)

print(Solution().minCostClimbingStairs([10,15,20]))
print(Solution().minCostClimbingStairs([1, 100, 1, 1, 1, 100, 1, 1, 100, 1]))
